### ISSUE TITLE
MQTT 5.0 cannot subscribe: Cannot parse topic

### ISSUE THREAD

**Issue by miqmago** [mqttjs/mqtt-connection#24]
I'm trying to use MQTT 5.0 but I'm always getting the same error: `Cannot parse topic`

```js
        client.connect({
            clientId: activeInterface.mac_address.replace(/:/ig, '').toUpperCase(),
            username: 'bla',
            password: 'bla',
            protocolId: 'MQTT',
            protocolVersion: 5,
        });

        client.on('connack', () => {
            console.log('client connected... sending subscribe');
            client.subscribe({
                messageId: 0,
                subscriptions: [{
                    topic: 'whatever',
                    qos: 0,
                    nl: false, // no Local MQTT 5.0 flag
                    rap: true, // Retain as Published MQTT 5.0 flag
                    rh: 1, // Retain Handling MQTT 5.0
                }],
            });
        });
```

I've add this line in [`parser.js:466`](https://github.com/mqttjs/mqtt-packet/blob/master/parser.js#L466) for debugging:

```js
console.log(length, end, this._list.length, this.packet.length, this._list.toString('utf8'));
```

It always fails (whatever topic I place):

```
2167 2172 13 13 '\u0000\u0000\u0000\bwhatever\u0000'
client error Error: Cannot parse topic
    at Parser._parseSubscribe (/node_modules/mqtt-connection/node_modules/mqtt-packet/parser.js:304:48)
    at Parser._parsePayload (/node_modules/mqtt-connection/node_modules/mqtt-packet/parser.js:99:14)
    at Parser.parse (/node_modules/mqtt-connection/node_modules/mqtt-packet/parser.js:39:41)
    at DestroyableTransform.process [as _transform] (/node_modules/mqtt-connection/lib/parseStream.js:14:17)
    at DestroyableTransform.Transform._read (/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at DestroyableTransform.Transform._write (/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:172:83)
    at doWrite (/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:428:64)
    at writeOrBuffer (/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:417:5)
    at DestroyableTransform.Writable.write (/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:334:11)
    at Socket.ondata (_stream_readable.js:639:20)
```

If I switch to MQTT lower version it works perfectly...

**Comment by miqmago** [comment]
If I place a console.log at `_parseNum` [parser.js:499](https://github.com/mqttjs/mqtt-packet/blob/master/parser.js#L499) it seems that _pos is one byte misfit:

```js
  console.log(this._list._bufs);
  console.log(this._list._bufs.map(b => b.toString()), this._pos, result);
```

```
[ <Buffer 00 00 00 08 77 68 61 74 65 76 65 72 00> ]
[ '\u0000\u0000\u0000\bwhatever\u0000' ] 3 2167
```

**Comment by miqmago** [comment]
Seems a problem of options:
When the packet is receiving connack, it calls `setOptions` with connectPacket. I don't see a way to return protocol version from server in `connack`, or the connack from client is not getting original options or assigning original options to packet. That way, `protocolVersion` is being lost.

I've ended by overriding options manually, doing this:

```js

    mqttConnection.on('connack', this.conAck.bind(this));

    conAck(resp) {
        this.mqttConnection.setOptions({
            protocolVersion: 5,
        });
        this.mqttConnection.subscribe({
            messageId: 1,
            subscriptions: [{
                topic: 'whatever',
                qos: 0,
                nl: false, // no Local MQTT 5.0 flag
                rap: true, // Retain as Published MQTT 5.0 flag
                rh: 1, // Retain Handling MQTT 5.0
            }],
        });
    }
```

**Comment by miqmago** [comment]
Problem reported at https://github.com/mqttjs/mqtt-packet/issues/50
